File: src/main/java/com/eightbitforest/termu/emu/nes/system/Nes.java
package main.java.com.eightbitforest.termu.emu.nes.system;

import main.java.com.eightbitforest.termu.emu.nes.rom.NesRom;

public class Nes {

    private NesRom cartridge;
    private Cpu cpu;
    private Ppu ppu;
    private Mem mem;
    private Clock clock;

    public void insertCartridge(NesRom rom) {
        this.cartridge = rom;
    }

    public void start() {
        // TODO: Thread this
        mem = new Mem(cartridge);
        cpu = new Cpu(mem);
        ppu = new Ppu(cpu, mem);
        clock = new Clock(cpu, ppu);

        new Thread(() -> cpu.start()).start();
        ppu.start();
//        new Thread(() -> ppu.start()).start();

//        clock.start();
    }

    public void stop() {
    }
}

File: src/emu/nes/system/nes.cpp
#include "nes.h"

nes::~nes() {
    delete cpu;
    delete ppu;
    delete mem;
}

void nes::insert_cartridge(nes_rom *rom) {
    this->cartridge = rom;
}

void nes::set_display(emu_display* display) {
	this->display = display;
}


void nes::start() {
	if (cartridge == nullptr) {
		throw emu_exception("A rom must be loaded before starting the emulator.");
	}

	if (display == nullptr) {
		throw emu_exception("A display must be set before starting the emulator.");
	}

    mem = new ::mem(cartridge);
    ppu = new ::ppu(mem, display);
    cpu = new ::cpu(mem, ppu);

    cpu->start();
}

void nes::stop() {
}
